### DISCUSSION TITLE
Is there a way to run tasks before or instead of "npm install"?

### DISCUSSION THREAD

**Discussion by jpgoelz** [renovatebot/renovate#25845] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.59.8

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Is there a possibility to prevent Renovate from updating and running `npm install` for specific dependencies so that we can instead call our own command as a replacement? Something like pre-update commands or a flag to just not run `npm install`?

**Background:**

Angular is very particular about the way it is updated and very strict with its dependency restrictions.
An upgrade to a major version of Angular requires upgrading a mix of major and minor dependencies (see [previous discussion](https://github.com/renovatebot/renovate/discussions/25831)), as well as upgrading TypeScript at the same time (there are no TS version overlaps, e.g. Angular 16 requires TS <5.2.x while Angular 17 requires >=5.2.x). It does not seem correct to group TS with Angular permanently.
This results in always broken Renovate merge requests regarding Angular due to version conflicts.

You are actually not supposed to upgrade Angular by running `ng update @angular/core` resp. `ng update @angular/cli` instead of changing the versions in package.json and then running `npm install`, 
The command will install the correct versions of all needed dependencies (including TypeScript) as well as running some migrations on changed configurations, etc.

### What you tried so far.

1. I have tried using `postUpgradeTasks`:
   The Renovate docs actually use Angular as the example for [`allowPostUpgradeCommandTemplating`](https://docs.renovatebot.com/self-hosted-configuration/#allowpostupgradecommandtemplating).
   However, it has never worked so far, because there is always some version conflict with some dependency, especially zone.js or TypeScript. Due to the conflict, the post upgrade commands cannot determine the proper executable to run.
   `--force` is also not an option, because `ng update` checks, if the new version is already installed in `node_modules` and exits without doing anything else, if it finds the target version ("Package '@angular/core' is already up to date.").

2. I have some grouping rules now, that at least allow me to [group major versions with some minor versions](https://github.com/renovatebot/renovate/discussions/25831). However, because I would not want to add TypeScript to that group, this has the same problem.

**Comment by jpgoelz** [comment]
I would also be happy with a solution, that does not perform an upgrade step but rather just create an issue, that there is an update available, so that I can then manually perform `ng update` and not clog our CI/CD pipelines with broken merge requests.

**Comment by rarkins** [comment, marked as answer]
There's only postUpgradeTasks, and a feature request for [preUpgradeTasks](https://github.com/renovatebot/renovate/issues/8804)

**Reply by jpgoelz** [reply]
I searched for any feature requests with numerous search queries and did not find it. Or are you considering this as the mentioned feature request (because you linked to this discussion 😅)?

**Reply by rarkins** [reply]
Link updated. Should be https://github.com/renovatebot/renovate/issues/8804

**Reply by jpgoelz** [reply]
Oh, I searched only for discussions, not issues. My bad. 🤦
